src/o365/spfx/commands/project/project-externalize/rules/DynamicRule.ts

```diff
@@ -0,0 +1,77 @@
+import { BasicDependencyRule } from "./BasicDependencyRule";
+import { Project } from "../../project-upgrade/model";
+import { ExternalizeEntry } from "../model/ExternalizeEntry";
+import * as fs from 'fs';
+import request from '../../../../../../request';
+
+
+export class DynamicRule extends BasicDependencyRule {
+  public async visit(project: Project): Promise<ExternalizeEntry[]> {
+    if(project.packageJson) {
+      const validPackageNames = Object.getOwnPropertyNames(project.packageJson.dependencies)
+                            .filter(x => x.indexOf(this.typesPrefix) === -1)
+                            .filter(x => this.restrictedModules.indexOf(x) === -1);
+
+      return (await Promise.all(validPackageNames.map((x) => this.getExternalEntryForPackage(x, project))))
+              .filter(x => x !== undefined).map(x => x as ExternalizeEntry);
+    } else {
+      return [];
+    }
+  }
+  private getExternalEntryForPackage = async (packageName: string, project: Project): Promise<ExternalizeEntry | undefined> => {
+    const version = project.packageJson && project.packageJson.dependencies[packageName];
+    const filePath = this.cleanFilePath(this.getFilePath(packageName));
+    if(version && filePath) {
+      let url = this.getFileUrl(packageName, version, filePath);
+      let testResult = await this.testUrl(url);
+      if (!url.endsWith('.min.js')) {
+        const minUrl = url.replace('.js', '.min.js');
+        const minResult = await this.testUrl(minUrl);
+        if(minResult) {
+          url = minUrl;
+          testResult = true;
+        }
+      }
+      if(testResult) {
+        return {
```

Comment:
Won't this lead to incorrect results as it's not taking into account if the particular package is AMD or non-AMD?

Reply:
which part specifically? the globalName?
This was one of my earliest comments, I guess we should test the module type but I'm not sure how to best implement that. Maybe if you could share some insights one how you did it for the Rencore tool?

Reply:
Let me check if there is something we could use.